lib/spack/spack/solver/asp.py

```diff
@@ -3340,7 +3343,41 @@ def variant_value(self, node, name, value):
 
         self._specs[node].update_variant_validate(name, value)
 
+    def _associated_branch(self, version_string, package_class):
+        if hasattr(package_class, "git"):
+            version = vn.Version(version_string)
+            version_dict = package_class.versions.get(version, {})
+            return version_dict.get("branch", None)
+        else:
+            return None
+
+    def _retrieve_latest_git_hash(self, branch_name, package_class):
+        # remote git operations can sometimes have banners so we must parse the output for a sha
+        query = (
+            spack.util.git.git(required=True)(
+                "ls-remote", "-h", package_class.git, branch_name, output=str, error=str
+            )
+            .strip()
+            .split()
+        )
+        sha = [hunk for hunk in query if bool(COMMIT_VERSION.match(hunk))]
+        if len(sha) == 1:
+            return sha[0]
+        else:
+            raise InternalConcretizerError(
+                (
+                    f"Failure to fetch git sha from {package_class.git} --branch {branch_name}. "
+                    "Confirm network via staging this package, and post a bug report if that "
+                    "succeeds."
+                )
+            )
```

Comment:
I think there's a simpler way to validate that all of the git usage and output parsing has worked. I think this covers all possible failures, without needing to check the COMMIT_HASH regex against sections that shouldn't match it.

```suggestion
        try:
            query = spack.util.git.git(required=True)(
                "ls-remote", "-h", package_class.git, branch_name, output=str, error=str
            )
            sha, ref = query.strip().split()
            assert COMMIT_VERSION.match(sha)
        except (ProcessError, ValueError, AssertionError):
            raise InternalConcretizerError(
                (
                    f"Failure to fetch git sha from {package_class.git} --branch {branch_name}. "
                    "Confirm network via staging this package, and post a bug report if that "
                    "succeeds."
                )
            )
```

We could add in `assert ref = f"refs/heads/{branch_name}` if we want to preserve way the current code ensures the 2nd returned value doesn't match COMMIT_VERSION. I also think there's a cleaner error message we could write, but after two bad stabs at it I'm not coming up with something right now.

Reply:
Also not a fan of the message so let's keep iterating on it. This change fails for legal banners like the one that inspired this change:

```
$ git ls-remote -h [git address]
Legal Notice Reminder - SF 2902-NTU (2-2009)

         WARNING NOTICE TO USERS: SNL Systems Are Monitored

 This Sandia National Laboratories (SNL) information system is monitored.
 It is the property of the U.S. Government and is for authorized uses only.

 --------------- Be aware you have no privacy on these systems. --------------

    All users of this system and all information on this system are
    subject to interception, monitoring, recording, copying, auditing,
    and inspection at the discretion of Sandia Corporation, and
    disclosure to Sandia Corporation and third parties, including,
    but not limited to, the United States Government,any authorized
    investigative and law enforcement personnel, and officials of
    other entities, both domestic and foreign. By using or accessing
    this system, you consent to and permit all of the above without
    limit as to when such action may be undertaken.

    Unauthorized or improper use of this SNL system may result in
    disciplinary, administrative, civil, and/or criminal penalties.

    LOGIN to acknowledge and agree with these terms and conditions or
   DISCONTINUE all efforts to access or use this SNL information system.

[sha]        refs/heads/master
```
I'm trying out some things to see if I can simplify it, but I'm not certain if things like this will always get piped the same way/same ordering. The current implementation is robust, but not the nicest code for sure.

Reply:
I tested this some more and the legal notice/header is piped to stderrr. So we could just send that to `os.devnull` for this instance. I suspect that will always be the case since the command is from git and I don't think you can hijack the stdout. I'll update on that assumption.